spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingSyntaxErrorListener.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2022-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.jpa.repository.query;
+
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+
+/**
+ * A {@link BaseErrorListener} that will throw a {@link QueryParsingSyntaxError} if the query is invalid.
+ *
+ * @author Greg Turnquist
+ * @since 3.1
+ */
+class QueryParsingSyntaxErrorListener extends BaseErrorListener {
```

Comment:
We could turn this into an enum or create a singleton to avoid instance creation on each usage.

Reply:
We could, unless we want to pass in the original query. This then becomes a factory operation. I can poke at it to see what's possible.